### PR TITLE
Allow configurable managed identity when launching Azure clusters

### PR DESCRIPTION
**Pull request by mjd3**
## Why are these changes needed?
Currently, one is not able to use a previously configured managed identity when launching Azure clusters. This can cause issues with trying to connect multiple clusters to shared resources such as storage accounts or key vaults that lie outside the resource group of the created VMs as they won't able to access those shared resources. It also makes it much easier to tear down clusters, as the entire resource group can be deleted without fear of deleting any shared resources between clusters.

This change allows the user to pass the name of an existing managed identity and its resource group as part of the provider options in the cluster yaml file. If those fields are provided, a new managed identity is not created and the existing one is used. This change is backwards-compatible since if the new fields are not provided, the default behavior is the same as before - create a new managed identity in the same resource group as the deployment. I've verified both cases locally by launching a cluster with/without these parameters specified and viewing the (created) MSIs in the portal.

## Related issue number
Discussed in https://github.com/ray-project/ray/issues/29215, but does not appear that it was ever PRed.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [x] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by mjd3** [comment]
@ericl @architkulkarni @hongchaodeng let me know if there's anything else I need to do for this PR; I don't believe I can assign anyone to review it.

**Comment by architkulkarni** [comment]
Hi @gramhagen, would you mind reviewing this PR if you have time?

**Comment by architkulkarni** [comment]
@mjd3 would it make sense to add documentation here? https://docs.ray.io/en/latest/cluster/vms/references/ray-cluster-configuration.html#cluster-configuration-provider-type

**Review by gramhagen — APPROVED** [review]
looks good, adding an example declaration in the config would be helfpul
https://github.com/ray-project/ray/blob/master/python/ray/autoscaler/azure/example-full.yaml

**Comment by mjd3** [comment]
Thanks @gramhagen and @architkulkarni for the feedback; I added to both of those locations with examples!

File: doc/source/cluster/vms/references/ray-cluster-configuration.rst [hunk 1]
```diff
@@ -129,6 +129,8 @@ Provider
             :ref:`location <cluster-configuration-location>`: str
             :ref:`resource_group <cluster-configuration-resource-group>`: str
             :ref:`subscription_id <cluster-configuration-subscription-id>`: str
+            :ref:`msi_name <cluster-configuration-msi-name>`: str
+            :ref:`msi_resource_group <cluster-configuration-msi-resource-group>`: str
```
Comment (architkulkarni) [inline doc/source/cluster/vms/references/ray-cluster-configuration.rst:133]: @mjd3 I think you'll need to also add the corresponding sections in the doc and write descriptions of the fields, do you mind double checking this?

Comment (architkulkarni) [reply doc/source/cluster/vms/references/ray-cluster-configuration.rst:133]: E.g. https://docs.ray.io/en/latest/cluster/vms/references/ray-cluster-configuration.html#provider-subscription-id

Comment (mjd3) [reply doc/source/cluster/vms/references/ray-cluster-configuration.rst:133]: Oops thanks for the catch! I _think_ I updated them correctly this time; let me know if something is missing still.

**Comment by architkulkarni** [comment]
Looks good! https://anyscale-ray--43897.com.readthedocs.build/en/43897/cluster/vms/references/ray-cluster-configuration.html#provider

**Review by architkulkarni — APPROVED** [review]

**Comment by anyscalesam** [comment]
@mjd3 would love to learn more about your use case if you'd be game to share; Slack me on Ray Slack @anyscalesam  if you have a chance!

**Pull request merged**